gpcontrib/arenadata_toolkit/src/track_files.c

```diff
@@ -0,0 +1,1228 @@
+		v_stmt.kind = VAR_RESET;
+		v_stmt.args = NIL;
+	}
+	else
+	{
+		v_stmt.kind = VAR_SET_VALUE;
+		v_stmt.args = list_make1(&arg);
+		elog(LOG, "[arenadata_toolkit] setting relkinds %s in database %u for tracking", buf.data, dbid);
+	}
+
+	tf_guc_unlock_relkinds_once();
+
+	AlterDatabaseSet(&stmt);
+	pfree(buf.data);
+
+	PG_RETURN_BOOL(true);
+}
+
+static bool
+is_valid_relstorage(char relstorage)
+{
+	return (relstorage == 'h' || relstorage == 'a' || relstorage == 'c' ||
+			relstorage == 'x' || relstorage == 'v' || relstorage == 'f');
+}
+
+Datum
+tracking_set_relstorages(PG_FUNCTION_ARGS)
+{
+	char	   *relstorages_str = NameStr(*PG_GETARG_NAME(0));
+	Oid			dbid = PG_GETARG_OID(1);
+	char	   *token;
+	char	   *str_copy;
+	bool		seen_relstorages[256] = {false};
+	StringInfoData buf;
+	AlterDatabaseSetStmt stmt;
+	VariableSetStmt v_stmt;
+	A_Const		arg;
+
+	dbid = dbid == InvalidOid ? MyDatabaseId : dbid;
+
+	initStringInfo(&buf);
+	str_copy = pstrdup(relstorages_str);
+	token = strtok(str_copy, ",");
+	while (token != NULL)
+	{
+		if (strlen(token) != 1 || !is_valid_relstorage(token[0]))
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("Invalid relstorage type: %s", token),
+			errhint("Valid relstorages are: 'h', 'x', 'a', 'v', 'c', 'f'")));
+
+		if (!seen_relstorages[(unsigned char)token[0]])
+		{
+			appendStringInfoChar(&buf, token[0]);
+			appendStringInfoChar(&buf, ',');
+			seen_relstorages[(unsigned char)token[0]] = true;
+		}
+		token = strtok(NULL, ",");
+	}
+	pfree(str_copy);
+
+	stmt.type = T_AlterDatabaseSetStmt;
+	stmt.dbname = get_database_name(dbid);
+
+	if (stmt.dbname == NULL)
+		elog(ERROR, "[arenadata_toolkit] database %u does not exist", dbid);
+
+	stmt.setstmt = &v_stmt;
+
+	v_stmt.type = T_VariableSetStmt;
+	v_stmt.name = "arenadata_toolkit.tracking_relstorages";
+	v_stmt.is_local = false;
+
+	arg.type = T_A_Const;
+	arg.val.type = T_String;
+	arg.val.val.str = buf.data;
+	arg.location = -1;
+
+	if (buf.len > 0 && buf.data[buf.len - 1] == ',')
+	{
+		buf.data[buf.len - 1] = '\0';
+		buf.len--;
+	}
+
+	if (buf.len == 0)
+	{
+		v_stmt.kind = VAR_RESET;
+		v_stmt.args = NIL;
+	}
+	else
+	{
+		v_stmt.kind = VAR_SET_VALUE;
+		v_stmt.args = list_make1(&arg);
+		elog(LOG, "[arenadata_toolkit] setting relstorages %s in database %u for tracking", buf.data, dbid);
+	}
+
+	tf_guc_unlock_relstorages_once();
+
+	AlterDatabaseSet(&stmt);
+
+	pfree(buf.data);
+
+	PG_RETURN_BOOL(true);
+}
+
+Datum
+tracking_trigger_initial_snapshot(PG_FUNCTION_ARGS)
+{
+	Oid			dbid = PG_GETARG_OID(0);
+
+	dbid = dbid == InvalidOid ? MyDatabaseId : dbid;
+	elog(LOG, "[arenadata_toolkit] tracking_trigger_initial_snapshot dbid: %d", dbid);
+
+	if (!bloom_set_trigger_bits(&tf_shared_state->bloom_set, dbid, true))
+		elog(ERROR, "Failed to find corresponding filter to database %u", dbid);
+
+	if (Gp_role == GP_ROLE_DISPATCH)
+	{
+		char	   *cmd = psprintf("select arenadata_toolkit.tracking_trigger_initial_snapshot(%d)", dbid);
```

Comment:
Some format strings use `%d` for `dbid`, some use `%u`. `Oid` is unsigned, maybe change relevant format modifiers to `%u`?